### DISCUSSION TITLE
Open Watcom V2 CMake build

### DISCUSSION THREAD

**Discussion by FredEckert** [open-watcom/open-watcom-v2#716] [Q&A] [answered]
Hello :-)

Does anyone have an example of a cmake toolchain file for cross compiling with open watcom v2? Host: Window10 x64, Target: DOS 32-Bit DOS4G/W.

I have scoured the web and studied the cmake docs but I am having trouble making the leap to a clear understanding of how to construct such a toolchain file. I have tried all sorts of things but can't get cmake to successfully build code that builds just fine with wmake and a good old fashioned wmake [makefile](https://github.com/open-watcom/open-watcom-v2/blob/master/bld/src/goodies/makefile.c).

Side Note: I am really seeking an out-of-source build system for open watcom v2. cmake seems to be the correct approach for this. If there is any other recommend approach, I would be interested in learning about it.

**Comment by jmalak** [comment, marked as answer]
From CMake version 3.18 OpenWatcom Cross-compile support for DOS, OS/2, Windows and Linux is included in CMake.
Details you can see in #687 #599 #594 

Below are cross-compilation commands for CMake 3.18 and above (run on Linux or Windows). It supposes configured OW for your host (Windows or Linux).

16-bit DOS Target cmake command
`cmake -G "Watcom WMake" -D CMAKE_SYSTEM_NAME=DOS -D CMAKE_SYSTEM_PROCESSOR=I86 ...`

32-bit DOS Target cmake command
`cmake -G "Watcom WMake" -D CMAKE_SYSTEM_NAME=DOS ...`

16-bit OS2 Target cmake command
`cmake -G "Watcom WMake" -D CMAKE_SYSTEM_NAME=OS2 -D CMAKE_SYSTEM_PROCESSOR=I86 ...`

32-bit OS2 Target cmake command
`cmake -G "Watcom WMake" -D CMAKE_SYSTEM_NAME=OS2 ...`

32-bit Windows Target cmake command
`cmake -G "Watcom WMake" -D CMAKE_SYSTEM_NAME=Windows ...`

32-bit Linux Target cmake command
`cmake -G "Watcom WMake" -D CMAKE_SYSTEM_NAME=Linux ...`

**Reply by jmalak** [reply]
If you are using CMake GUI then use following setup.
![image](https://user-images.githubusercontent.com/2017260/126758340-c6e252cf-1796-4c5e-b77e-6b359e1aaa25.png)

Sample for 16-bit DOS
![image](https://user-images.githubusercontent.com/2017260/126758530-b5b01142-f229-463e-aa55-f5e57d3c1f03.png)

Sample for 32-bit DOS
![image](https://user-images.githubusercontent.com/2017260/126758759-c5996618-e4ca-4c8d-befe-004e805d5284.png)

etc...
Command line definition of target system and processor architecture is used for appropriate GUI fields.
Default processor architecture is 32-bit that need not be specified.

**Reply by FredEckert** [reply]
Yes. This is what I have tried but is not working for me. I am launching the Open Watcom V2 "Open Watcom Build Environment" command prompt from the installer created start menu icon. This action sets up my open watcom environment. I don't have CMake in my path so I type the full path:

```
C:\TEST>"c:\Program Files\CMake\bin\cmake.exe" -G "Watcom WMake" -D CMAKE_SYSTEM_NAME=DOS -S . -B build
```

If I am trying the gui, I type:
```
C:\TEST>"C:\Program Files\CMake\bin\cmake-gui.exe"
```

I do this because I believe that cmake needs to have knowledge of open watcom environment variables...

To build the code I do this:
```
C:\TEST>"C:\Program Files\CMake\bin\cmake.exe" --build build
```

The build always fails.

If you see anything obviously wrong with the process above, please let me know. I will try making a smaller test case to see if I can get something simple working...

**Reply by jmalak** [reply]
I think it should work.
What exactly failed?
Creation of build make files or build process when you run wmake?

**Reply by FredEckert** [reply]
The creation process succeeds like this:

> C:\TEST>"c:\Program Files\CMake\bin\cmake.exe" -G "Watcom WMake" -D CMAKE_SYSTEM_NAME=DOS -S . -B build
> -- The C compiler identification is OpenWatcom 2.0
> -- The CXX compiler identification is OpenWatcom 2.0
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Check for working C compiler: C:/WATCOM/binnt64/wcl386.exe - skipped
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Check for working CXX compiler: C:/WATCOM/binnt64/wcl386.exe - skipped
> -- Configuring done
> -- Generating done
> -- Build files have been written to: C:/TEST/build

The build process fails like this:

> C:\TEST>"C:\Program Files\CMake\bin\cmake.exe" --build build
> Open Watcom Make Version 2.0 beta Feb  4 2021 00:39:12 (64-bit)
> Copyright (c) 2002-2021 The Open Watcom Contributors. All Rights Reserved.
> Portions Copyright (c) 1988-2002 Sybase, Inc. All Rights Reserved.
> Source code is available under the Sybase Open Watcom Public License.
> See http://www.openwatcom.org/ for details.
> Scanning dependencies of target ex22
> [ 20%] Building C object src/CMakeFiles/ex22.dir/DBG.H.obj
> [ 40%] Building C object src/CMakeFiles/ex22.dir/EX22.H.obj
> [ 60%] Building C object src/CMakeFiles/ex22.dir/EX22.C.obj
> [ 80%] Building C object src/CMakeFiles/ex22.dir/EX22_MAIN.C.obj
> C:\TEST\src\EX22_MAIN.C(47): Error! E1077: Missing '}'
> C:\TEST\src\EX22_MAIN.C(23): Warning! W303: Parameter 'argv' has been defined, but not referenced
> C:\TEST\src\EX22_MAIN.C(23): Warning! W303: Parameter 'argc' has been defined, but not referenced
> C:\TEST\src\EX22_MAIN.C(47): Warning! W107: Missing return value for function 'main'
> C:\TEST\src\EX22_MAIN.C(48): Error! E1116: An id list not allowed except for function definition
> C:\TEST\src\EX22_MAIN.C(48): Warning! W304: Return type 'int' assumed for function 'scope_demo'
> C:\TEST\src\EX22_MAIN.C(48): Error! E1062: Inconsistent return type for function 'scope_demo'
> C:\TEST\src\EX22_MAIN.C(48): Note! N2003: source conversion type is 'int'
> C:\TEST\src\EX22_MAIN.C(48): Note! N2004: target conversion type is 'void'
> C:\TEST\src\EX22_MAIN.C(48): Note! N2002: 'scope_demo' defined in: C:\TEST\src\EX22_MAIN.C(6)
> C:\TEST\src\EX22_MAIN.C(49): Error! E1009: Expecting ',' but found '*'
> C:\TEST\src\EX22_MAIN.C(49): Error! E1116: An id list not allowed except for function definition
> C:\TEST\src\EX22_MAIN.C(49): Warning! W304: Return type 'int' assumed for function 'scope_demo'
> C:\TEST\src\EX22_MAIN.C(49): Error! E1062: Inconsistent return type for function 'scope_demo'
> C:\TEST\src\EX22_MAIN.C(49): Note! N2003: source conversion type is 'int'
> C:\TEST\src\EX22_MAIN.C(49): Note! N2004: target conversion type is 'void'
> C:\TEST\src\EX22_MAIN.C(49): Note! N2002: 'scope_demo' defined in: C:\TEST\src\EX22_MAIN.C(6)
> C:\TEST\src\EX22_MAIN.C(51): Error! E1009: Expecting ')' but found '('
> C:\TEST\src\EX22_MAIN.C(51): Error! E1009: Expecting ')' but found '('
> C:\TEST\src\EX22_MAIN.C(51): Error! E1094: Function cannot return a function
> C:\TEST\src\EX22_MAIN.C(51): Error! E1094: Function cannot return a function
> C:\TEST\src\EX22_MAIN.C(51): Warning! W304: Return type 'int' assumed for function 'fprintf'
> C:\TEST\src\EX22_MAIN.C(51): Error! E1062: Inconsistent return type for function 'fprintf'
> C:\TEST\src\EX22_MAIN.C(51): Note! N2003: source conversion type is 'int (void)(void))()(void))()'
> C:\TEST\src\EX22_MAIN.C(51): Note! N2004: target conversion type is 'int'
> C:\TEST\src\EX22_MAIN.C(51): Note! N2002: 'fprintf' defined in: C:\WATCOM\h\stdio.h(345)
> C:\TEST\src\EX22_MAIN.C(51): Error! E1151: Parameter count does not agree with previous definition
> C:\TEST\src\EX22_MAIN.C(51): Error! E1044: Variable 'fprintf' has incomplete type
> C:\TEST\src\EX22_MAIN.C(51): Error! E1026: Invalid declarator
> C:\TEST\src\EX22_MAIN.C(51): Error! E1009: Expecting ',' but found '&'
> C:\TEST\src\EX22_MAIN.C(51): Error! E1024: Declared symbol '__iob' is not in parameter list
> C:\TEST\src\EX22_MAIN.C(51): Error! E1009: Expecting ',' but found ')'
> C:\TEST\src\EX22_MAIN.C(51): Error! E1026: Invalid declarator
> C:\TEST\src\EX22_MAIN.C(51): Error! E1026: Invalid declarator
> C:\TEST\src\EX22_MAIN.C(51): Error! E1009: Expecting ',' but found '[INFO] (%s:%d:) '
> C:\TEST\src\EX22_MAIN.C(51): Error! E1147: Too many errors: compilation aborted
> Error: Compiler returned a bad status compiling "C:\TEST\src\EX22_MAIN.C"
> Error(E42): Last command making (src\CMakeFiles\ex22.dir\EX22_MAIN.C.obj) returned a bad status
> Error(E02): Make execution terminated
> Error(E42): Last command making (src\CMakeFiles\ex22.dir\all) returned a bad status
> Error(E02): Make execution terminated
> Error(E42): Last command making (all) returned a bad status
> Error(E02): Make execution terminated

Perhaps the problem is in my CMakeLists.txt files...

Top level file:

```
# CMakeList.txt : Top-level CMake project file, do global configuration
# and include sub-projects here.
#
cmake_minimum_required(VERSION 3.21)
project("ex22" VERSION 0.0.0.1)

# Include sub-projects.
add_subdirectory("src")

```
Lower level file:

```
# CMakeList.txt : CMake project for ex22, include source and define
# project specific logic here.
#

# Add source to this project's executable.


################################################################################
# Source groups
################################################################################
set(Header_Files
    "DBG.H"
    "EX22.H"
)
source_group("Header Files" FILES ${Header_Files})

set(Source_Files
    "EX22.C"
    "EX22_MAIN.C"
)
source_group("Source Files" FILES ${Source_Files})

set(ALL_FILES
    ${Header_Files}
    ${Source_Files}
)

set_source_files_properties(${ALL_FILES} PROPERTIES LANGUAGE C)


################################################################################
# Target
################################################################################
add_executable(${PROJECT_NAME} ${ALL_FILES})


################################################################################
# Dependencies
################################################################################


# TODO: Add tests and install targets if needed.

```

I attached two test case files: TEST-cmake.zip and TEST-makefile.zip. They are both the same code. The traditional in-source-build makefile build works. The cmake fails to build. 

[TEST-cmake.zip](https://github.com/open-watcom/open-watcom-v2/files/6870322/TEST-cmake.zip)
[TEST-makefile.zip](https://github.com/open-watcom/open-watcom-v2/files/6870324/TEST-makefile.zip)

**Reply by jmalak** [reply]
it looks like a problem with warning level and your source code required standard. 
The main problem is that your code  is C99 standard, but OW compiler use C89/90 standard by default.
Default OW compiler warning level is 1.
Cmake make file uses warning level 3 (more strict) and C89/90 standard.

Anyway cmake is working OK, only problem with compiler options used by Cmake definition, need correction.
You need to add extra option for C99 standard -za99 to cmake and set warning level to 1 if you want the same behavior.

**Reply by FredEckert** [reply]
I added the -za99 compiler flag:
```
# CMakeList.txt : CMake project for ex22, include source and define
# project specific logic here.
#

# Add source to this project's executable.

################################################################################
# Source groups
################################################################################
set(Header_Files
    "DBG.H"
    "EX22.H"
)
source_group("Header Files" FILES ${Header_Files})

set(Source_Files
    "EX22.C"
    "EX22_MAIN.C"
)
source_group("Source Files" FILES ${Source_Files})

set(ALL_FILES
    ${Header_Files}
    ${Source_Files}
)

set_source_files_properties(${ALL_FILES} PROPERTIES LANGUAGE C)
set(CMAKE_C_FLAGS "-za99")

################################################################################
# Target
################################################################################
add_executable(${PROJECT_NAME} ${ALL_FILES})


################################################################################
# Dependencies
################################################################################


# TODO: Add tests and install targets if needed.

```
and it worked:

> C:\TEST>"C:\Program Files\CMake\bin\cmake.exe" --build build
> Open Watcom Make Version 2.0 beta Feb  4 2021 00:39:12 (64-bit)
> Copyright (c) 2002-2021 The Open Watcom Contributors. All Rights Reserved.
> Portions Copyright (c) 1988-2002 Sybase, Inc. All Rights Reserved.
> Source code is available under the Sybase Open Watcom Public License.
> See http://www.openwatcom.org/ for details.
> Scanning dependencies of target ex22
> [ 20%] Building C object src/CMakeFiles/ex22.dir/DBG.H.obj
> [ 40%] Building C object src/CMakeFiles/ex22.dir/EX22.H.obj
> [ 60%] Building C object src/CMakeFiles/ex22.dir/EX22.C.obj
> [ 80%] Building C object src/CMakeFiles/ex22.dir/EX22_MAIN.C.obj
> [100%] Linking C executable ex22.exe
> [100%] Built target ex22
> 
> C:\TEST>

Thank You!

**Comment by jmalak** [comment]
New page in OW 2.0 Wiki was created [OW tools usage with CMake](https://github.com/open-watcom/open-watcom-v2/wiki/OW-tools-usage-with-CMake)
